### DISCUSSION TITLE
Multiple loaders yet only one schema class

### DISCUSSION THREAD

**Discussion by smolinari** [Nikaple/nest-typed-config#19] [Q&A] [answered]
Hi,

I'm about to try it out, but I figured I'd ask first. 

If I use multiple loaders, and say I have nested objects for two sets of config data, one set is environment variables (from the one loader) and one from a database (from another loader), will `nest-typed-config` know to "fill in" the right object properties automatically?

I'd imagine so, but the docs aren't clear about it.

Scott

**Comment by Nikaple** [comment, marked as answer]
Your config is recursively merged, check out the [source code](https://github.com/Nikaple/nest-typed-config/blob/main/lib/typed-config.module.ts#L82).

**Reply by smolinari** [reply]
Ok. Thanks for the reply. Trying it out now. 

Would it also make sense to be able to offer a schema class per loader and then merge them? 

Scott

**Reply by Nikaple** [reply]
You can call `forRoot` multiple times. For example:
```ts
@Module({
  imports: [
    TypedConfigModule.forRoot({ schema: Schema1, load: fileLoader() }),
    TypedConfigModule.forRoot({ schema: Schema2, load: envLoader() }),
  ]
})
```

PS: Please do not share any class or sub-class between Schema1 and Schema2, or Nest.js won't know which class to inject.

**Reply by Nikaple** [reply]
I've created a draft PR #20 , you can take a look if you have further questions :smiley_cat:

**Reply by smolinari** [reply]
Thanks. @Nikaple. Appreciate the help. One last open challenge I have is creating a loader that would use a database service to get config information. Is this even possible? I'm not seeing how to get a service injected as the loader. 

Scott

**Reply by Nikaple** [reply]
That's not possible. As the most fundamental module of your application, config module should not rely on any other modules, and that's why we didn't even provide an option to inject other services. But, however, you can define your own "config-from-database" providers with Nest's custom providers functionality. Such as:

```ts
import { DatabaseModule, DatabaseService } from './database.module';

export class ConfigFromDatabaseSchema {
  someConfig: string;
}

@Module({
  imports: [DatabaseModule],
  providers: [
    {
      provide: ConfigFromDatabaseSchema,
      useFactory(databaseService) {
        // validation is omitted for simplicity
        return databaseService.query('select * from config limit 1')
      },
      inject: [DatabaseService]
    }
  ],
})
export class DatabaseConfigModule {}
```

Then you can inject your "config-from-database" with `ConfigFromDatabaseSchema`, please refer to [Nest's documentation](https://docs.nestjs.com/fundamentals/custom-providers) for more information.

**Reply by smolinari** [reply]
Hey. Thanks so much @Nikaple. I understand the config being fundamental part and the direction you suggest is where I was heading too. 

Appreciate the support. Keep up the good work!

Scott

**Reply by smolinari** [reply]
@Nikaple - Just wanted to let you know, like your suggestion above, I put together a factory provider to generate the config from the database. Works like a charm. Thanks for your help and your package (which I'm still using for the env config). 

Scott

**Reply by Nikaple** [reply]
Glad that it works, you're welcome :smile_cat:
